File: flutter_app/lib/api/skills_api.dart
import 'api_client.dart';

class SkillsApi {
  SkillsApi(this._client);

  final ApiClient _client;

  Future<List<Map<String, dynamic>>> listSkills() async {
    final data = await _client.get('/skills/list') as List<dynamic>;
    return data.map((dynamic e) => Map<String, dynamic>.from(e as Map)).toList();
  }

  Future<List<Map<String, dynamic>>> searchTeach({
    String? query,
    int? skillId,
    String? level,
    String? mode,
  }) async {
    final params = <String, String>{};
    if (query != null && query.isNotEmpty) {
      params['q'] = query;
    }
    if (skillId != null) {
      params['skill_id'] = '$skillId';
    }
    if (level != null && level.isNotEmpty) {
      params['level'] = level;
    }
    if (mode != null && mode.isNotEmpty) {
      params['mode'] = mode;
    }

    final data = await _client.get('/teach/search', query: params) as List<dynamic>;
    return data.map((dynamic e) => Map<String, dynamic>.from(e as Map)).toList();
  }

  Future<void> addTeach({
    required int skillId,
    required String level,
    required String mode,
    String? description,
  }) async {
    await _client.post('/teach/add', body: {
      'skill_id': skillId,
      'level': level,
      'mode': mode,
      'description': description,
    });
  }

  Future<void> addLearn({
    required int skillId,
    required String levelNeeded,
    String? notes,
  }) async {
    await _client.post('/learn/add', body: {
      'skill_id': skillId,
      'level_needed': levelNeeded,
      'notes': notes,
    });
  }
}

File: frontend/src/api/skillsApi.ts
import { asArray, asRecord } from '../models/parse';
import { ApiClient } from './apiClient';

export type SkillRow = Record<string, unknown>;

export class SkillsApi {
  constructor(private readonly client: ApiClient) {}

  async listSkills(): Promise<SkillRow[]> {
    const data = asArray(await this.client.get('/skills/list'));
    return data.map(item => asRecord(item));
  }

  async searchTeach(filters?: {
    query?: string;
    skillId?: number;
    level?: string;
    mode?: string;
  }): Promise<SkillRow[]> {
    const params: Record<string, string> = {};
    if (filters?.query !== undefined && filters.query.trim() !== '') {
      params.q = filters.query.trim();
    }
    if (filters?.skillId !== undefined) {
      params.skill_id = String(filters.skillId);
    }
    if (filters?.level !== undefined && filters.level.trim() !== '') {
      params.level = filters.level.trim();
    }
    if (filters?.mode !== undefined && filters.mode.trim() !== '') {
      params.mode = filters.mode.trim();
    }

    const data = asArray(await this.client.get('/teach/search', params));
    return data.map(item => asRecord(item));
  }

  async addTeach(payload: {
    skillId: number;
    level: string;
    mode: string;
    description?: string;
  }): Promise<void> {
    await this.client.post('/teach/add', {
      skill_id: payload.skillId,
      level: payload.level,
      mode: payload.mode,
      description: payload.description,
    });
  }

  async addLearn(payload: {
    skillId: number;
    levelNeeded: string;
    notes?: string;
  }): Promise<void> {
    await this.client.post('/learn/add', {
      skill_id: payload.skillId,
      level_needed: payload.levelNeeded,
      notes: payload.notes,
    });
  }
}
